src/Components/ResourcePicker/ResourcePicker.tsx

```diff
@@ -0,0 +1,649 @@
+                case AzureResourceTypes.StorageBlobContainer:
+                    return t('resourcesPicker.errors.invalidContainerUrl');
+                default:
+                    return undefined;
+            }
+        }
+    }, [selectedOption, resourceType, isValidUrlStr, t]);
+
+    const isValuesEqual = (value1: string, value2: string) => {
+        if (displayField === AzureResourceDisplayFields.url) {
+            if (value1?.endsWith('/')) {
+                value1 = value1.slice(0, -1);
+            }
+            if (value2?.endsWith('/')) {
+                value2 = value2.slice(0, -1);
+            }
+            return value1 === value2;
+        } else {
+            return value1 === value2;
+        }
+    };
+
+    const handleOnChange = useCallback(
+        (option, value) => {
+            if (option) {
+                setSelectedOption(option);
+                setSelectedKey(option.data?.id ?? option.text);
+                if (onResourceChange) {
+                    onResourceChange(
+                        option.data || option.text,
+                        options
+                            ?.filter(
+                                (option) =>
+                                    option.itemType !==
+                                    SelectableOptionMenuItemType.Header
+                            )
+                            .map((option) => option.data || option.text) || []
+                    );
+                }
+            } else {
+                const newParsedOptionValue = parsedOptionText(value);
+
+                const existingOption = options.find((option) =>
+                    isValuesEqual(option.text, newParsedOptionValue)
+                );
+                if (!existingOption) {
+                    const newOption = {
+                        key: newParsedOptionValue,
+                        text: newParsedOptionValue,
+                        styles: comboBoxOptionStyles
+                    };
+
+                    let newOptions = deepCopy(options);
+                    if (
+                        displayField !== AzureResourceDisplayFields.url ||
+                        (displayField === AzureResourceDisplayFields.url &&
+                            isValidUrlStr(newParsedOptionValue))
+                    ) {
+                        const freeFromOptionsHeaderExist = options?.find(
+                            (option) =>
+                                option.itemType ===
+                                    SelectableOptionMenuItemType.Header &&
+                                option.text === freeformOptionsHeaderText
+                        );
+                        newOptions = freeFromOptionsHeaderExist
+                            ? [...newOptions, newOption]
+                            : [...newOptions, freeformOptionsHeader, newOption];
+                        setOptions(newOptions);
```

Comment:
the fact that we need to set state multiple times in a row screams that we should be using a reducer for this component.